packages/builders/src/interactions/modals/Modal.ts

```diff
@@ -53,49 +61,103 @@ export class ModalBuilder implements JSONEncodable<APIModalInteractionResponseCa
 	 * @param customId - The custom id to use
 	 */
 	public setCustomId(customId: string) {
-		this.data.custom_id = customIdValidator.parse(customId);
+		this.data.custom_id = customId;
 		return this;
 	}
 
 	/**
-	 * Adds components to this modal.
+	 * Adds action rows to this modal.
 	 *
 	 * @param components - The components to add
 	 */
-	public addComponents(
+	public addActionRows(
 		...components: RestOrArray<
-			ActionRowBuilder<ModalActionRowComponentBuilder> | APIActionRowComponent<APIModalActionRowComponent>
+			| ActionRowBuilder
+			| APIActionRowComponent<APIModalActionRowComponent>
+			| ((builder: ActionRowBuilder) => ActionRowBuilder)
 		>
 	) {
-		this.components.push(
-			...normalizeArray(components).map((component) =>
-				component instanceof ActionRowBuilder
-					? component
-					: new ActionRowBuilder<ModalActionRowComponentBuilder>(component),
-			),
-		);
+		const normalized = normalizeArray(components);
+		const resolved = normalized.map((row) => resolveBuilder(row, ActionRowBuilder));
+
+		this.data.components.push(...resolved);
+
 		return this;
 	}
 
 	/**
-	 * Sets components for this modal.
+	 * Sets the action rows for this modal.
 	 *
 	 * @param components - The components to set
 	 */
-	public setComponents(...components: RestOrArray<ActionRowBuilder<ModalActionRowComponentBuilder>>) {
-		this.components.splice(0, this.components.length, ...normalizeArray(components));
+	public setActionRows(...components: RestOrArray<ActionRowBuilder>) {
+		this.data.components.splice(0, this.data.components.length, ...normalizeArray(components));
```

Comment:
This can use `this.spliceActionRows()` now, which should make the parameter of this accept more types